Feature/non mandatory keystore encryption

dakingha69 opened the pull request:
Closes #257 

- `wallet.createAccount()` do not load and encrypt the created wallet, instead only returns the created `TLWallet`
- `wallet.loadAccount()` do not require a serialized encrypted wallet as a parameter, instead expect a `TLWallet` -> no wallet decryption required
- `wallet.recoverFromSeed()` / `wallet.recoverFromPrivateKey()` do not require a serialized encrypted wallet as a parameter, instead only the mnemonic / private key -> no wallet decryption required
- `wallet.recoverFromEncryptedWallet()` as additional recovery method
- `wallet.encryptWallet()` as additional method to encrypt wallet with password

compojoom commented on src/User.ts:
```diff
@@ -125,6 +102,25 @@ export class User {
+    return this.wallet.encryptWallet(
+      tlWallet,
+      typeof password === 'string' ? password : this.defaultPassword,
+      typeof password === 'function' ? password : progressCallback
```
Are you sure about all this?

typeof password === 'string' => typeof '' === 'string' - do we want to use empty string as password?

the next line typeof password === 'function' did you mean progressCallback?

dakingha69 replied:
> typeof password === 'string' => typeof '' === 'string' - do we want to use empty string as password?

It is possible with ethers js, so I would also allow it.

> the next line typeof password === 'function' did you mean progressCallback?

Wanted to give user the option to specify a callback without a password like
```js
user.encryptWallet(tlWallet, progressCallback => {...})
```
But not sure about this...

berndbohmeier commented on src/typings.ts:
```diff
@@ -420,17 +420,46 @@ export interface Signature {
+export interface EncryptedIdentityWalletSchema extends TLWalletSchema {
+  meta: {
+    ethersKeystore: string
+    identityAddress: string
```
I would remove the identityAddress here, we can just recalculate the address here, because it is now deterministic.

berndbohmeier commented on src/wallets/EthersWallet.ts:
```diff
@@ -129,10 +140,12 @@ export class EthersWallet implements TLWallet {
     if (!('TLWalletVersion' in deserializedWallet)) {
-      // Use the old serialising method if no TL_WALLET_VERSION key in the deserialized wallet
-      encryptedKeystore = serializedWallet
+      // Use the old serializing method if no TL_WALLET_VERSION key in the deserialized wallet
```
If you want to be able to load the old format for backwards compatibility, it should be in loadAccount, otherwise I would remove it.

dakingha69 replied:
Ok so it would be okay for you if I omit backwards compatibility for now? So I would keep version 1 of the TLWallet but include breaking changes as we have no real user currently anyways.

berndbohmeier commented on src/wallets/EthersWallet.ts:
```diff
@@ -89,34 +72,62 @@ export class EthersWallet implements TLWallet {
+   * @param password Password to encrypt wallet with.
+   * @param progressCallback Optional encryption progress callback.
+   */
+  public async encryptWallet(
```
I understood it that we made this function to return the standard format, so we should just return the encrypted keystore here.

dakingha69 replied:
changed it so that it only returns the string as returned by `ethers.Wallet.encrypt`

berndbohmeier commented on src/typings.ts:
```diff
@@ -420,17 +420,46 @@ export interface Signature {
+}
+
+interface EthersWalletMeta {
+  walletFromEthers: ethers.Wallet
```
I thought we aggreed on saving the SigningKey, which consists of the mnemonic, the derivation path and the private key, now?

berndbohmeier replied:
I would not like to expose a type which is not from us, this breaks abstraction.

dakingha69 replied:
Misunderstood something. Will changes this

dakingha69 commented on src/wallets/EthersWallet.ts:
```diff
@@ -53,26 +49,13 @@ export class EthersWallet implements TLWallet {
-      address: this.address,
-      pubKey: this.pubKey,
-      serializedWallet
+      address: await walletFromEthers.getAddress(),
```
ok will change this

dakingha69 commented on src/wallets/EthersWallet.ts:
```diff
@@ -89,34 +72,62 @@ export class EthersWallet implements TLWallet {
+   * @param ethersWallet `TLWallet` of type `WALLET_TYPE_ETHERS`.
+   */
+  public async loadAccount(ethersWallet: EthersWalletSchema): Promise<void> {
+    this.walletFromEthers = ethersWallet.meta.walletFromEthers
```
now added to `loadAccount`. it was in `recoverFromEncryptedJson` before

berndbohmeier commented on src/typings.ts:
```diff
@@ -420,17 +420,35 @@ export interface Signature {
-  walletType: string
+export interface IdentityWalletSchema extends TLWalletSchema {
+  type: WalletTypeIdentity
+  meta: EthersWalletMeta
```
I would here also store the address, so that you dont have to recalculate the address.

berndbohmeier commented on src/typings.ts:
```diff
@@ -420,17 +420,35 @@ export interface Signature {
+export interface SigningKey {
+  privateKey: string
+  mnemonic: string
+  derivationPath: string
```
Thinking about it, we dont have to store the derivation path, because we currently also do not support using in in recover from mnemonic. So we can just hard code it as we do it in recover from mnemonic.

berndbohmeier commented on src/utils.ts:
```diff
@@ -389,6 +393,33 @@ export const trimUrl = (url: string): string => {
+): ethers.Wallet {
+  const { signingKey } = wallet.meta
+  return signingKey.mnemonic && signingKey.derivationPath
+    ? ethers.Wallet.fromMnemonic(signingKey.mnemonic, signingKey.derivationPath)
```
Isnt that too slow? isnt the private key in that case recalculated from the mnemonic? We should always recover from the private key directly and additionally restore the mnemonic.

berndbohmeier commented on src/wallets/EthersWallet.ts:
```diff
@@ -53,27 +54,11 @@ export class EthersWallet implements TLWallet {
-      serializedWallet
-    }
+  public async createAccount(): Promise<UserObject<EthersWalletSchema>> {
+    const walletFromEthers = ethers.Wallet.createRandom()
```
Where do you store the wallet?

berndbohmeier commented on src/wallets/EthersWallet.ts:
```diff
@@ -89,121 +74,77 @@ export class EthersWallet implements TLWallet {
+    privateKey: string
+  ): Promise<UserObject<EthersWalletSchema>> {
+    const walletFromEthers = new ethers.Wallet(privateKey)
+    return this.getAccountWithEthersWallet(walletFromEthers)
```
where do you store the wallet?

berndbohmeier commented on src/wallets/IdentityWallet.ts:
```diff
@@ -150,128 +128,86 @@ export class IdentityWallet implements TLWallet {
-      typeof progressCallback === 'function' && progressCallback
+    const walletFromEthers = utils.getWalletFromEthers(identityWallet)
+    this.walletFromEthers = walletFromEthers
+    this.identityAddress = calculateIdentityAddress(
```
I would not recalculate the address here, because that might be slow. 
Lets just store it.

berndbohmeier commented on tests/Fixtures.ts:
```diff
@@ -39,27 +42,16 @@ export const parametrizedTLNetworkConfig = [
+  return Promise.all(
+    tlInstances.map(async tl => {
+      const createdUser = await tl.user.create()
+      await tl.user.load(createdUser.wallet)
```
So you split up the creation and loading of an account on purpose? Why? I think it is intuitive that create account, restore account will not change the state, but load will.

dakingha69 commented:
I changed the naming into `credentials`. So we have now smth like `wallet.createCredentials`, `wallet.recoverCredentialsFrom*`, `wallet.loadCredentials`, etc. I also removed the type `UserObject` as this was unnecessary and misleading. Therefore I added the property `address` directly to the type `TLWalletCredentials`.

berndbohmeier commented on tests/unit/User.test.ts:
```diff
@@ -73,16 +71,13 @@ describe('unit', () => {
-        assert.isString(loadedUser.address)
-        assert.isString(loadedUser.serializedWallet)
-        assert.isString(loadedUser.pubKey)
+        const loadedUser = await user.loadFrom(TL_WALLET_DATA)
```
this should rather test that it is loaded into the library, so something like checking the address would be good.

berndbohmeier commented:
Can be merged after you addressed the small problems if you need it, but there are still performance problems. 
But we should try to fix this one last issue, that we do everything twice. 
So we generate the ethers wallet and then take the data from it to generate it again. This is slow.

berndbohmeier commented on src/wallets/WalletFromEthers.ts:
```diff
@@ -0,0 +1,84 @@
+
+  public static createRandom() {
+    const { privateKey, mnemonic } = super.createRandom()
+    return new this(privateKey, mnemonic)
```
But I checked, instantiation should be fast, so probably does not matter.

dakingha69 commented on package.json:
```diff
@@ -13,8 +13,8 @@
     "test:e2e": "nyc mocha --timeout 30000 --require isomorphic-fetch -r ts-node/register tests/e2e/**.test.ts",
-    "test:integration": "nyc mocha --timeout 5000 --require isomorphic-fetch -r ts-node/register tests/integration/**.test.ts",
-    "test:unit": "nyc mocha --timeout 5000 --require isomorphic-fetch -r ts-node/register tests/unit/**.test.ts",
+    "test:integration": "nyc mocha --timeout 30000 --require isomorphic-fetch -r ts-node/register tests/integration/**.test.ts",
```
ups no only unit tests
